src/properties.jl

```diff
@@ -0,0 +1,668 @@
+
+- `alloc_time`: the timing for the allocation of storage space for a dataset's
+  raw data; one of:
+   - `:default`
+
+   - `:early`: allocate all space when the dataset is created
+
+   - `:incremental`: Allocate space incrementally, as data is  written to the
+     dataset
+
+   - `:late`: Allocate all space when data is first written to the dataset.
+
+  See $(h5doc("H5P_SET_ALLOC_TIME")).
+
+- `chunk`: a tuple containing the size of the chunks to store each dimension.
+  See $(h5doc("H5P_SET_CHUNK")) (note that this uses Julia's column-major
+  ordering).
+
+- `external`: A tuple of `(name,offset,size)`, See $(h5doc("H5P_SET_EXTERNAL")).
+
+- `filters` (only valid when `layout=:chunked`): a filter or vector of filters
+  that are applied to applied to each chunk of a dataset, see [Filters](@ref).
+  When accessed, will return a [`Filters.FilterPipeline`](@ref) object that can
+  be modified in-place.
+
+- `layout`: the type of storage used to store the raw data for a dataset. Can be
+  one of:
+
+   - `:compact`: Store raw data in the dataset object header in file. This
+     should only be used for datasets with small amounts of raw data.
+
+   - `:contiguous`: Store raw data separately from the object header in one
+     large chunk in the file.
+
+   - `:chunked`: Store raw data separately from the object header as chunks of
+     data in separate locations in the file.
+
+   - `:virtual`:  Draw raw data from multiple datasets in different files.
+
+  See $(h5doc("H5P_SET_LAYOUT")).
+
+
+The following options are shortcuts for the various filters, and are set-only.
+They will be appended to the filter pipeline in the order in which they appear
+
+- `blosc = true | level`: set the [`Filters.BloscFilter`](@ref) compression
+  filter; argument can be either `true`, or the compression level.
+
+- `deflate = true | level`: set the [`Filters.Deflate`](@ref) compression
+  filter; argument can be either `true`, or the compression level.
+
+- `fletcher32 = true`: set the [`Filters.Fletcher32`](@ref) checksum filter.
+
+- `shuffle = true`: set the [`Filters.Shuffle`](@ref) filter.
+
+"""
+@propertyclass DatasetCreateProperties API.H5P_DATASET_CREATE
+superclass(::Type{DatasetCreateProperties}) = ObjectCreateProperties
+
+@enum_property(alloc_time,
+    :default     => API.H5D_ALLOC_TIME_DEFAULT,
+    :early       => API.H5D_ALLOC_TIME_EARLY,
+    :incremental => API.H5D_ALLOC_TIME_INCR,
+    :late        => API.H5D_ALLOC_TIME_LATE)
+
+# reverse indices
+function get_chunk(p::Properties)
+    dims, N = API.h5p_get_chunk(p)
+    ntuple(i -> Int(dims[N-i+1]), N)
+end
+set_chunk!(p::Properties, dims) = API.h5p_set_chunk(p, length(dims), API.hsize_t[reverse(dims)...])
+
+@enum_property(layout,
+    :compact    => API.H5D_COMPACT,
+    :contiguous => API.H5D_CONTIGUOUS,
+    :chunked    => API.H5D_CHUNKED,
+    :virtual    => API.H5D_VIRTUAL)
+
+
+# filters getters/setters
+get_filters(p::Properties) = Filters.FilterPipeline(p)
```

Comment:
These filters getters and setters should be defined Properties not DatasetCreateProperties ?

Reply:
We could be more specific in some cases, but since we can't represent the inheritance with subtypes, it isn't possible to define them all this way, e.g. getters/setters defined on `ObjectCreateProperties` wouldn't work on `FileCreateProperties`.

In the end, since the getters and setters are really an internal API, I figured it was more effort than it was worth.

Reply:
Oh right all the getters/setters are defined on the abstract type. If they weren't then it would mess up in the inheritance chain and not find them. Ok I think understand this then.

Reply:
Lastly, wouldn't calling `get_filters` always reset the FilterPipeline?

Reply:
By reset I mean creat a new `FilterPipeline` object thus dropping the existing pipeline?

Reply:
Oh nvm I now understand what this is doing.

Reply:
Yeah, `FilterPipeline` is just a wrapper object.